table/merging_iterator.cc

```diff
@@ -188,6 +212,20 @@ class MergingIterator : public InternalIterator {
       minHeap_.pop();
     }
     current_ = CurrentForward();
+
+    if (range_del_agg_ != nullptr && Valid()) {
+      // When we cross over to a new range tombstone, we use an optimization to
+      // seek the iterators whose keys are all covered by the range tombstone.
+      // TODO(ajkr): this must be optimized to seek only when tombstone changed.
+      auto optimized_seek_endpoint = range_del_agg_->GetEndpoint(
+          key(), RangeDelPositioningMode::kForwardTraversal);
```

Comment:
What do you think about making it zero function call overhead only in case of `ReadOptions::ignore_range_deletions == true` (the default is false)?

> building Merging iterator, if range_del_agg_ is empty, not put it to merging iterator and leave it empty, in this case, we don't need to go into this code path if people don't use range delete.

It can be the case that files opened later during the scan contain range deletions. Let me see if we can make this function inlined, then the overhead can be a simple check for emptiness (via a flag?), and such a branch would be predicted 100% correctly for someone who doesn't use `DeleteRange()`. It seems it should be possible though I haven't investigated yet.

Reply:
I will revisit this today.